File: DataStructureAndAlgorithm/Graph/Kruskals/C++/Kruskals.cpp
#include "Kruskals.h"
#include <iostream>

DisjointSet::DisjointSet(int n) {
    parent.resize(n);
    for (int i = 0; i < n; ++i) {
        parent[i] = i;
    }
}

int DisjointSet::find(int i) {
    if (parent[i] == i) {
        return i;
    }
    return parent[i] = find(parent[i]);
}

void DisjointSet::unite(int i, int j) {
    int root_i = find(i);
    int root_j = find(j);
    if (root_i != root_j) {
        parent[root_i] = root_j;
    }
}

void kruskalMST(int numVertices, std::vector<Edge>& edges) {
    std::vector<Edge> result;
    int i = 0; // Index used to pick next edge
    int e = 0; // Index used for result[]

    // Sort all the edges in non-decreasing order of their weight.
    std::sort(edges.begin(), edges.end(), [](const Edge& a, const Edge& b) {
        return a.weight < b.weight;
    });

    DisjointSet ds(numVertices);

    while (e < numVertices - 1 && i < edges.size()) {
        Edge next_edge = edges[i++];

        int x = ds.find(next_edge.src);
        int y = ds.find(next_edge.dest);

        // If including this edge does't cause cycle, include it in result and increment the count of edges.
        if (x != y) {
            result.push_back(next_edge);
            ds.unite(x, y);
            e++;
        }
    }

    std::cout << "Following are the edges in the constructed MST" << std::endl;
    for (i = 0; i < e; ++i) {
        std::cout << result[i].src << " -- " << result[i].dest << " == " << result[i].weight << std::endl;
    }
}

File: DataStructureAndAlgorithm/Graph/Kruskals/Python/kruskals.py
class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n))

    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def union(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            self.parent[root_i] = root_j
            return True
        return False

def kruskal_mst(num_vertices, edges):
    # edges is a list of tuples: (weight, u, v)
    edges.sort()
    
    ds = DisjointSet(num_vertices)
    mst = []
    min_cost = 0

    for weight, u, v in edges:
        if ds.union(u, v):
            mst.append((u, v, weight))
            min_cost += weight

    print("Edges in the MST:")
    for u, v, weight in mst:
        print(f"{u} - {v} : {weight}")
    print(f"Minimum Cost Spanning Tree: {min_cost}")
    return mst
